core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/FastRdfsForwardChainingSail.java

```diff
@@ -0,0 +1,311 @@
+    Set<Resource> properties = new HashSet<>();
+    List<Statement> subPropertyOfStatements = new ArrayList<>();
+    List<Statement> rangeStatements = new ArrayList<>();
+    List<Statement> domainStatements = new ArrayList<>();
+
+
+    Map<Resource, Set<Resource>> calculatedTypes = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedProperties = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedRange = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedDomain = new HashMap<>();
+    private boolean sharedSchema;
+
+    void clearInferenceTables() {
+        subClassOfStatements = new ArrayList<>();
+        properties = new HashSet<>();
+        subPropertyOfStatements = new ArrayList<>();
+        rangeStatements = new ArrayList<>();
+        domainStatements = new ArrayList<>();
+        calculatedTypes = new HashMap<>();
+        calculatedProperties = new HashMap<>();
+        calculatedRange = new HashMap<>();
+        calculatedDomain = new HashMap<>();
+    }
+
+    void readLock(FastRdfsForwardChainingSailConnection connection) {
+
+        if (numberOfThreadsWaitingForWriteLock.get() > 0) {
+//            System.err.println("starve reads");
+        }
+
+        while (numberOfThreadsWaitingForWriteLock.get() > 0) {
+            Thread.yield();
+        }
+
+        if (connection.lockStamp != 0) {
+            throw new IllegalStateException("Connection already has a lock!");
+        }
+        connection.lockStamp = readWriteLock.readLock();
+//        System.err.println("readLock: " + connection.lockStamp);
+
+    }
+
+    void releaseLock(FastRdfsForwardChainingSailConnection connection) {
+        readWriteLock.unlock(connection.lockStamp);
+//        System.err.println("Released lock: " + connection.lockStamp);
+
+        connection.lockStamp = 0;
+    }
+
+
+    AtomicInteger numberOfThreadsWaitingForWriteLock = new AtomicInteger(0);
+
+    void upgradeLock(FastRdfsForwardChainingSailConnection connection) {
+
+////        System.err.println("Attempt writelock: "+connection.lockStamp);
+
+        numberOfThreadsWaitingForWriteLock.incrementAndGet();
+
+        try {
+           while(true) {
+                long l = readWriteLock.tryConvertToWriteLock(connection.lockStamp);
+
+                if (l != 0) {
+                    long temp = connection.lockStamp;
+                    connection.lockStamp = l;
+                    if (temp != l) {
+//                        System.err.println("readLock: " + temp + " writeLock: " + connection.lockStamp);
+                    }
+
+
+                    return;
+                }
+
+                try {
+                    Thread.sleep(random.nextInt(2));
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+
+                // detect potential deadlock scenario
+                if(numberOfThreadsWaitingForWriteLock.get() > 1  //More than 1 connection waiting for a write lock
```

Comment:
Read/write locks allow multiple threads to hold a read lock, but if a thread holds a write lock no other thread can hold any kind of lock. Read locks are required for everything, and write locks are required only when modifying (or potentially modifying) the Tbox. 

Scenario one: Thread A and B both have read locks, and then thread B needs a write lock. It starves all new read locks (blocks threads acquiring read locks), then waits for them to be released and then acquires the write lock and continues. 

Scenario two: Thread A and B both have read locks, and then both threads need write locks. E.g. up until now they have been modifying the Abox, but now they both want to add Tbox triples. For thread A to acquire the write lock, thread B needs to release the read lock, and for thread B to acquire the write lock it needs thread A to release it's read lock. So the algorithm will randomly select one of the threads, release it's read lock and then throw an exception.